File: day1/day1_2.py
delimiters = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
word_to_num = { 'on': 1, 'tw' : 2, 'th' : 3, 'fo' : 4, 'fi' : 5, 'si' : 6, 'se' : 7, 'ei' : 8, 'ni' : 9 }

import time

with open('input.txt', 'r') as file:
    sum = 0
    
    for line in file.readlines():
        left_idx = 100000
        right_idx = -1

        # Find the leftmost and the rightmost numbers (numerically or verbally)
        for delim in (d for d in delimiters if d in line): 
            left_idx = min(left_idx, line.index(delim))
            right_idx = max(right_idx, line.rindex(delim))
                
        left_digit = 0
        right_digit = 0

        if line[left_idx].isdigit():
            left_digit = int(line[left_idx])
        else:
            left_digit = word_to_num[line[left_idx:left_idx+2]]

        if line[right_idx].isdigit():
            right_digit = int(line[right_idx])
        else:
            right_digit = word_to_num[line[right_idx:right_idx+2]]

        sum += left_digit * 10 + right_digit       
        
    print(sum)

File: day1/day1_2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <array>
#include <unordered_map>

int main()
{
    std::ifstream file("input.txt");

    std::vector<std::string> lines{};

    std::unordered_map<std::string, uint32_t> word_to_num{
        {"one", 1U},
        {"two", 2U},
        {"three", 3U},
        {"four", 4U},
        {"five", 5U},
        {"six", 6U},
        {"seven", 7U},
        {"eight", 8U},
        {"nine", 9U},
    };

    std::string _line{};
    while (std::getline(file, _line))
        lines.emplace_back(_line);
   
    size_t sum{};

    for (auto& line : lines)
    {
        // 1. Find all positions of one, two, three, four...
        std::array<std::vector<size_t>, 9> positions{};
        
        for (const auto& [word, num] : word_to_num)
        {
            size_t j{}, pos{};
            while ((pos = line.find(word, j)) != std::string::npos)
            { 
                positions[num-1U].push_back(pos);
                j = pos + word.size(); // + 1
            }
        }

        // 2. Replace all occurences
        for (uint_fast32_t i = 0U; i < 9U; ++i)
            for (const auto& pos : positions[i])
                line[pos + 1] = '0' + (char)i + 1;

        // 3. Find first and last
        char left = line[line.find_first_of("123456789")];
        char right = line[line.find_last_of("123456789")];

        sum += (size_t)((left - '0') * 10 + (right - '0'));
    }
   
    std::cout << sum << '\n';
}